crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,333 @@
+            id,
+            graph: self.clone(),
+        }
+    }
+
+    fn add_dependency(&self, src: NodeId, dst: NodeId) {
+        if let Some(system) = self.nodes.borrow_mut().get_mut(&dst) {
+            match system {
+                SystemDescriptor::Parallel(descriptor) => descriptor.after.push(src.dyn_clone()),
+                SystemDescriptor::Exclusive(descriptor) => descriptor.after.push(src.dyn_clone()),
+            }
+        } else {
+            panic!(
+                "Attempted to add dependency for {:?}, which doesn't exist.",
+                dst
+            );
+        }
+    }
+}
+
+/// A draining conversion to [SystemSet]. All other clones of the same graph will be empty
+/// afterwards.
+///
+/// [SystemSet]: crate::schedule::SystemSet
+impl From<SystemGraph> for SystemSet {
+    fn from(graph: SystemGraph) -> Self {
+        let mut system_set = SystemSet::new();
+        for (_, system) in graph.nodes.borrow_mut().drain() {
+            system_set = system_set.with_system(system);
+        }
+        system_set
+    }
+}
+
+#[derive(Clone)]
+pub struct SystemGraphNode {
+    id: NodeId,
+    graph: SystemGraph,
+}
+
+impl SystemGraphNode {
+    /// Creates a new node in the graph and adds the current node as it's dependency.
+    ///
+    /// This function can be called multiple times to add mulitple systems to the graph,
+    /// all of which will not execute until original node's system has finished running.
+    pub fn then(&self, next: impl Into<SystemDescriptor>) -> SystemGraphNode {
+        let node = self.graph.create_node(next.into());
+        self.graph.add_dependency(self.id, node.id);
+        node
+    }
+
+    /// Fans out from the given node into multiple dependent systems. All provided
+    /// systems will not run until the original node's system finishes running.
+    ///
+    /// Functionally equivalent to calling `SystemGraphNode::then` multiple times.
+    pub fn split_into<T: SystemGroup>(&self, system_group: T) -> T::Output {
+        system_group.split_into(self)
+    }
+}
+
+pub trait SystemGroup {
+    type Output;
+    fn split_into(self, src: &SystemGraphNode) -> Self::Output;
+}
+
+impl<T: Into<SystemDescriptor>> SystemGroup for Vec<T> {
+    type Output = Vec<SystemGraphNode>;
+    fn split_into(self, src: &SystemGraphNode) -> Self::Output {
+        self.into_iter().map(|sys| src.then(sys.into())).collect()
+    }
+}
+
+macro_rules! ignore_first {
+    ($_first:ident, $second:ty) => {
+        $second
+    };
+}
+
+macro_rules! impl_system_group_tuple {
+    ($($param: ident),*) => {
+        impl<$($param: Into<SystemDescriptor>),*> SystemGroup for ($($param,)*) {
+            // HACK: using repeat macros without using the param in it fails to compile.
+            // The ignore_first here "uses" the parameter and just discards it.
+            type Output = ($(ignore_first!($param, SystemGraphNode),)*);
+
+            #[inline]
+            #[allow(non_snake_case)]
```

Comment:
What's the `[allow(non_snake_case)]` suppressing?

Reply:
Line 219 destructures the tuple into variables using the tuple type parameters (T1, T2, T3...), which trigger a snake case warning. This suppresses it. If you look at the tuple implementation for SystemParam, it does the exact same thing.